hclsyntax/expression_test.go

```diff
@@ -1937,26 +1959,26 @@ func TestExpressionErrorMessages(t *testing.T) {
 			"The true and false result expressions must have consistent types. The 'false' value includes object attribute \"b\", which is absent in the 'true' value.",
 		},
 		{
-			"true ? listOf1Tuple : listOf0Tuple",
+			"true ? listOf2Tuple : listOf1Tuple",
 			&hcl.EvalContext{
 				Variables: map[string]cty.Value{
+					"listOf2Tuple": cty.ListVal([]cty.Value{cty.TupleVal([]cty.Value{cty.True, cty.Zero})}),
 					"listOf1Tuple": cty.ListVal([]cty.Value{cty.TupleVal([]cty.Value{cty.True})}),
-					"listOf0Tuple": cty.ListVal([]cty.Value{cty.EmptyTupleVal}),
```

Comment:
I am surprised that you were able to make this test fail by changing it in this way. An attempt to recreate this in a `terraform console` session:

```
> true ? tolist([[1]]) : tolist([[2,3]])
tolist([
  tolist([
    1,
  ]),
])

> false ? tolist([[1]]) : tolist([[2,3]])
tolist([
  tolist([
    2,
    3,
  ]),
])
>
```

Neither of these error. Am I missing something?

Reply:
I think the key here is that the two elements of the "2-tuple" must be of types that have no common base type, or else HCL will notice that it can convert the two tuples into lists of numbers to find a common type.

```
> true ? tolist([[1]]) : tolist([[2, true]])
╷
│ Error: Inconsistent conditional result types
│ 
│   on <console-input> line 1:
│   (source code not available)
│ 
│ The true and false result expressions must have consistent types.
│ Mismatched list element types: The 'true' tuple has length 1, but
│ the 'false' tuple has length 2.
╵
```

Reply:
Oh, that makes sense!

It might be worth leaving a comment to that effect in this test, then. Or renaming the variables to `tupleBoolNumber` and `tupleBool`.

Reply:
Yes, what Martin said!

I've added a comment explaining why the test cases fail.